cmd/snap-seccomp/main_test.go

```diff
@@ -148,21 +153,232 @@ func nativeEndian() binary.ByteOrder {
+		return rc;
+
+	execv(argv[2], (char *const *)&argv[2]);
+	perror("execv failed");
+	return 1;
+}
+`)
+
+var seccompSyscallRunnerContent = []byte(`
+#define _GNU_SOURCE
+#include<unistd.h>
+#include<sys/syscall.h>
+#include<stdlib.h>
+int main(int argc, char **argv) {
+   int l[7];
+   for (int i=0; i<7;i++)
+      l[i] = atoi(argv[i+1]);
+   // There might be architecture-specific requirements. see "man syscall"
+   // for details.
+   syscall(l[0], l[1], l[2], l[3], l[4], l[5], l[6]);
+   syscall(SYS_exit, 0, 0, 0, 0, 0, 0);
+}
+`)
+
+func lastKmsg() string {
+	output, err := exec.Command("dmesg").CombinedOutput()
+	if err != nil {
+		return err.Error()
+	}
+	l := strings.Split(string(output), "\n")
+	return fmt.Sprintf("Showing last 10 lines of dmesg:\n%s", strings.Join(l[len(l)-10:], "\n"))
+}
+
+func (s *snapSeccompSuite) SetUpSuite(c *C) {
+	// FIXME: we currently use a fork of x/net/bpf because of:
+	//   https://github.com/golang/go/issues/20556
+	// switch to x/net/bpf once we can simulate seccomp bpf there
+	bpf.VmEndianness = nativeEndian()
+
+	// build seccomp-load helper
+	s.seccompBpfLoader = filepath.Join(c.MkDir(), "seccomp_bpf_loader")
+	err := ioutil.WriteFile(s.seccompBpfLoader+".c", seccompBpfLoaderContent, 0644)
+	c.Assert(err, IsNil)
+	cmd := exec.Command("gcc", "-Werror", "-Wall", s.seccompBpfLoader+".c", "-o", s.seccompBpfLoader)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err = cmd.Run()
+	c.Assert(err, IsNil)
+
+	// build syscall-runner helper
+	s.seccompSyscallRunner = filepath.Join(c.MkDir(), "seccomp_syscall_runner")
+	err = ioutil.WriteFile(s.seccompSyscallRunner+".c", seccompSyscallRunnerContent, 0644)
+	c.Assert(err, IsNil)
+	cmd = exec.Command("gcc", "-Werror", "-Wall", "-static", s.seccompSyscallRunner+".c", "-o", s.seccompSyscallRunner, "-Wl,-static", "-static-libgcc")
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err = cmd.Run()
+	c.Assert(err, IsNil)
+}
+
+func (s *snapSeccompSuite) runBpfInKernel(c *C, seccompWhitelist, bpfInput string, expected int) {
+	// Common syscalls we need to allow for a minimal statically linked
+	// c program.
+	//
+	// If we compile a test program for each test we can get aways with
+	// a even smaller set of syscalls: execve,exit essentially. But it
+	// means a much longer test run (30s vs 2s). Commit d288d89 contains
+	// the code for this.
+	common := `
+execve
+uname
+brk
+arch_prctl
+readlink
+access
+sysinfo
+exit
+# i386
+set_thread_area
+`
+	bpfPath := filepath.Join(c.MkDir(), "bpf")
+	err := main.Compile([]byte(common+seccompWhitelist), bpfPath)
+	c.Assert(err, IsNil)
+
+	// syscallName;arch;arg1,arg2...
+	l := strings.Split(bpfInput, ";")
+	if len(l) > 1 && l[1] != "native" {
+		c.Logf("cannot use non-native in runBpfInKernel")
+		return
+	}
+	// Note that we will need to also skip: fadvise64_64,
+	//   ftruncate64, posix_fadvise, pread64, pwrite64, readahead,
+	//   sync_file_range, and truncate64.
+	// Once we start using those. See `man syscall`
```

Comment:
I updated it a bit to clarify it some more. It is essentially about that we cannot test some syscalls using the in-kernel VM because when we run them things explodes.